src/dawn/SIR/ASTNodeIterator.h

```diff
@@ -0,0 +1,308 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#ifndef DAWN_SIR_ASTASTNODEITERATOR_H
+#define DAWN_SIR_ASTASTNODEITERATOR_H
+
+#include "ASTStmt.h"
+#include "dawn/Support/IteratorRange.h"
+#include <list>
+#include <type_traits>
+
+namespace dawn {
+
+/// @brief (Abstract) Pre-order visiting iterator through the Stmts of one or more ASTs.
+/// Specializations should specify on which ASTs to iterate and in which order.
+/// @ingroup sir
+template <typename GeneratorType, bool onlyFirstLevel, typename Enable = void>
+class ASTNodeIterator {
+public:
+  virtual ~ASTNodeIterator() = 0;
+};
+
+#define ASTNodeIterator_Stmt                                                                       \
```

Comment:
I see now, that my confusion (comments below) are coming from these defines (it took me a while)
I would recommend to not use pragmas/defines for the coding, the obfuscate more than help with the readability

Reply:
Got it. I will remove the `ASTNodeIterator_Stmt` define and use the expanded version `ASTNodeIterator<ASTNode, onlyFirstLevel, typename std::enable_if<std::is_base_of<Stmt, ASTNode>::value>::type>`.

Reply:
Got it. I will remove the `ASTNodeIterator_Stmt` define and use the expanded version `ASTNodeIterator<ASTNode, onlyFirstLevel, typename std::enable_if<std::is_base_of<Stmt, ASTNode>::value>::type>`.